File: services/umbra/src/handlers/task-executor.ts
import { OpenRouterClient } from '@umbra/shared';
import { Envelope, UmbraPayload, ModuleResult, SupportedLanguage } from '@umbra/shared';
import Logger from '@umbra/shared/src/utils/logger';

export class TaskExecutor {
  private logger: Logger;
  private openRouterClient: OpenRouterClient;

  constructor(openRouterClient: OpenRouterClient) {
    this.logger = new Logger('TaskExecutor');
    this.openRouterClient = openRouterClient;
  }

  async executeTask(envelope: Envelope<UmbraPayload>): Promise<ModuleResult<any>> {
    const { reqId, userId, lang, payload } = envelope;
    const startTime = Date.now();

    try {
      this.logger.debug('Executing task', {
        reqId,
        userId,
        action: payload.action,
        intent: payload.intent
      });

      let result: any;

      // Handle tasks that Umbra can execute directly
      switch (payload.intent) {
        case 'calculation':
          result = await this.performCalculation(payload.message || '', lang);
          break;
        case 'translation':
          result = await this.performTranslation(payload.message || '', lang);
          break;
        case 'clarification_needed':
          result = await this.requestClarification(envelope);
          break;
        default:
          return {
            reqId,
            status: 'error',
            error: {
              type: 'functional',
              code: 'UNSUPPORTED_TASK',
              message: `Cannot execute task with intent: ${payload.intent}`,
              retryable: false
            }
          };
      }

      const durationMs = Date.now() - startTime;

      this.logger.audit('Task executed successfully', userId, {
        reqId,
        intent: payload.intent,
        durationMs
      });

      return {
        reqId,
        status: 'success',
        data: result,
        audit: {
          module: 'umbra-executor',
          durationMs
        }
      };

    } catch (error) {
      const durationMs = Date.now() - startTime;
      
      this.logger.error('Task execution failed', {
        reqId,
        userId,
        intent: payload.intent,
        error: error.message
      });

      return {
        reqId,
        status: 'error',
        error: {
          type: 'technical',
          code: 'EXECUTION_ERROR',
          message: 'Task execution failed',
          retryable: true
        },
        audit: {
          module: 'umbra-executor',
          durationMs
        }
      };
    }
  }

  async requestClarification(envelope: Envelope<UmbraPayload>): Promise<ModuleResult<any>> {
    const { reqId, lang, payload } = envelope;
    
    const questions = this.getClarificationQuestions(payload.message || '', lang);
    
    return {
      reqId,
      status: 'needs_validation',
      data: {
        type: 'clarification_request',
        message: questions.join('\n'),
        options: [
          'Finance - Document processing',
          'VPS - System management', 
          'Business - Client management',
          'Production - Workflow creation',
          'Creator - Media generation',
          'Translation - Language conversion',
          'Calculation - Math operations'
        ]
      }
    };
  }

  private async performCalculation(expression: string, lang: SupportedLanguage): Promise<any> {
    try {
      // Safe mathematical expression evaluation
      const mathExpression = this.extractMathExpression(expression);
      
      if (!mathExpression) {
        // Use AI for complex math problems
        const prompt = `Solve this math problem and show the result: ${expression}`;
        const result = await this.openRouterClient.generateText(prompt, 'anthropic/claude-3-haiku');
        
        return {
          type: 'calculation',
          expression,
          result,
          method: 'ai'
        };
      }

      // Simple expression evaluation
      const result = this.evaluateExpression(mathExpression);
      
      return {
        type: 'calculation',
        expression: mathExpression,
        result,
        method: 'direct'
      };

    } catch (error) {
      throw new Error(`Calculation failed: ${error.message}`);
    }
  }

  private async performTranslation(text: string, targetLang: SupportedLanguage): Promise<any> {
    try {
      const translatedText = await this.openRouterClient.translateText(text, targetLang);
      
      return {
        type: 'translation',
        originalText: text,
        translatedText,
        targetLanguage: targetLang
      };

    } catch (error) {
      throw new Error(`Translation failed: ${error.message}`);
    }
  }

  private extractMathExpression(text: string): string | null {
    // Extract simple mathematical expressions
    const patterns = [
      /(\d+(?:\.\d+)?)\s*([\+\-\*\/])\s*(\d+(?:\.\d+)?)/g,
      /(\d+)\s*\+\s*(\d+)/g,
      /(\d+)\s*\-\s*(\d+)/g,
      /(\d+)\s*\*\s*(\d+)/g,
      /(\d+)\s*\/\s*(\d+)/g
    ];

    for (const pattern of patterns) {
      const match = text.match(pattern);
      if (match) {
        return match[0];
      }
    }

    return null;
  }

  private evaluateExpression(expression: string): number {
    // Safe evaluation of simple mathematical expressions
    const sanitized = expression.replace(/[^0-9+\-*/.() ]/g, '');
    
    try {
      // Use Function constructor for safer evaluation than eval
      const result = new Function('return ' + sanitized)();
      
      if (typeof result !== 'number' || !isFinite(result)) {
        throw new Error('Invalid result');
      }
      
      return Math.round(result * 100) / 100; // Round to 2 decimal places
    } catch (error) {
      throw new Error('Invalid mathematical expression');
    }
  }

  private getClarificationQuestions(message: string, lang: SupportedLanguage): string[] {
    const questions = {
      EN: [
        "I understand you want help, but could you be more specific?",
        "Are you looking to:",
        "• Process a financial document (invoice, receipt)?",
        "• Check system status or run commands?",
        "• Manage client accounts?",
        "• Create a workflow or automation?",
        "• Generate media (images, videos, audio)?",
        "• Translate text?",
        "• Perform calculations?",
        "",
        "Please let me know which of these options matches what you need."
      ],
      FR: [
        "Je comprends que vous voulez de l'aide, mais pourriez-vous être plus précis ?",
        "Cherchez-vous à :",
        "• Traiter un document financier (facture, reçu) ?",
        "• Vérifier l'état du système ou exécuter des commandes ?",
        "• Gérer des comptes clients ?",
        "• Créer un workflow ou une automatisation ?",
        "• Générer des médias (images, vidéos, audio) ?",
        "• Traduire du texte ?",
        "• Effectuer des calculs ?",
        "",
        "Merci de me faire savoir laquelle de ces options correspond à vos besoins."
      ],
      PT: [
        "Entendo que você quer ajuda, mas poderia ser mais específico?",
        "Você está procurando:",
        "• Processar um documento financeiro (fatura, recibo)?",
        "• Verificar status do sistema ou executar comandos?",
        "• Gerenciar contas de clientes?",
        "• Criar um fluxo de trabalho ou automação?",
        "• Gerar mídia (imagens, vídeos, áudio)?",
        "• Traduzir texto?",
        "• Realizar cálculos?",
        "",
        "Por favor, me informe qual dessas opções corresponde ao que você precisa."
      ]
    };

    return questions[lang] || questions.EN;
  }

  /**
   * Handle simple text summarization
   */
  async summarizeText(text: string, lang: SupportedLanguage, maxLength: number = 500): Promise<string> {
    if (text.length <= maxLength) {
      return text; // No need to summarize
    }

    const prompt = `Summarize the following text in ${maxLength} characters or less. Keep it clear and concise:\n\n${text}`;
    
    return this.openRouterClient.generateText(prompt, 'anthropic/claude-3-haiku');
  }

  /**
   * Extract key information from user messages
   */
  extractParameters(message: string): Record<string, any> {
    const parameters: Record<string, any> = {};

    // Extract common parameters
    const patterns = {
      amount: /\$?(\d+(?:\.\d{2})?)/g,
      date: /(\d{1,2}[\/\-]\d{1,2}[\/\-]\d{2,4})/g,
      email: /([a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,})/g,
      phone: /(\+?1?[-.\s]?\(?[0-9]{3}\)?[-.\s]?[0-9]{3}[-.\s]?[0-9]{4})/g,
      url: /(https?:\/\/[^\s]+)/g
    };

    for (const [key, pattern] of Object.entries(patterns)) {
      const matches = message.match(pattern);
      if (matches) {
        parameters[key] = matches.length === 1 ? matches[0] : matches;
      }
    }

    return parameters;
  }

  /**
   * Check if task requires validation
   */
  requiresValidation(intent: string, parameters: Record<string, any>): boolean {
    const criticalIntents = [
      'vps_execute',
      'client_manage',
      'workflow_deploy'
    ];

    return criticalIntents.includes(intent) || 
           parameters.delete || 
           parameters.remove ||
           parameters.restart;
  }
}

File: python_services/umbra/umbra_service/handlers/task_executor.py
"""Task executor for simple tasks that can be handled by the main agent."""

import re
import json
from typing import Dict, Any, Optional
from datetime import datetime

from umbra_shared import (
    OpenRouterClient,
    UmbraLogger,
    Envelope,
    UmbraPayload,
    ModuleResult,
    SupportedLanguage,
)


class TaskExecutor:
    """Execute simple tasks directly in the main agent."""
    
    def __init__(self, openrouter_client: OpenRouterClient):
        self.logger = UmbraLogger("TaskExecutor")
        self.openrouter_client = openrouter_client
    
    async def execute_task(
        self,
        envelope: Envelope[UmbraPayload]
    ) -> ModuleResult[Dict[str, Any]]:
        """Execute a simple task based on the payload action."""
        start_time = datetime.utcnow()
        req_id = envelope.req_id
        action = envelope.payload.action
        message = envelope.payload.message or ""
        
        try:
            self.logger.info("Executing task",
                           req_id=req_id,
                           action=action)
            
            result_data = None
            
            if action == "execute":
                # Determine task type from message
                task_type = self._determine_task_type(message)
                
                if task_type == "calculate":
                    result_data = await self._handle_calculation(message, envelope.lang)
                elif task_type == "translate":
                    result_data = await self._handle_translation(message, envelope.lang)
                elif task_type == "help":
                    result_data = await self._handle_help(message, envelope.lang)
                else:
                    result_data = await self._handle_general_query(message, envelope.lang)
            
            elif action == "clarify":
                result_data = await self._handle_clarification(message, envelope.lang)
            
            else:
                result_data = {
                    "response": f"Unknown action: {action}",
                    "task_type": "error"
                }
            
            duration_ms = int((datetime.utcnow() - start_time).total_seconds() * 1000)
            
            self.logger.info("Task execution completed",
                           req_id=req_id,
                           task_type=result_data.get("task_type"))
            
            return ModuleResult(
                req_id=req_id,
                status="success",
                data=result_data,
                audit={
                    "module": "umbra-task-executor",
                    "duration_ms": duration_ms,
                    "task_type": result_data.get("task_type")
                }
            )
            
        except Exception as e:
            duration_ms = int((datetime.utcnow() - start_time).total_seconds() * 1000)
            
            self.logger.error("Task execution failed",
                            req_id=req_id,
                            action=action,
                            error=str(e))
            
            return ModuleResult(
                req_id=req_id,
                status="error",
                error={
                    "type": "technical",
                    "code": "EXECUTION_ERROR",
                    "message": f"Failed to execute task: {str(e)}",
                    "retryable": True
                },
                audit={
                    "module": "umbra-task-executor",
                    "duration_ms": duration_ms
                }
            )
    
    def _determine_task_type(self, message: str) -> str:
        """Determine the type of simple task from the message."""
        message_lower = message.lower()
        
        # Math calculations
        if re.search(r"\d+\s*[+\-*/]\s*\d+", message) or \
           any(word in message_lower for word in ["calculate", "compute", "math", "arithmetic"]):
            return "calculate"
        
        # Translation
        if any(word in message_lower for word in ["translate", "translation", "mean in", "say in"]):
            return "translate"
        
        # Help requests
        if any(word in message_lower for word in ["help", "assistance", "what can you", "how to"]):
            return "help"
        
        return "general"
    
    async def _handle_calculation(
        self,
        message: str,
        language: SupportedLanguage
    ) -> Dict[str, Any]:
        """Handle mathematical calculations."""
        try:
            # Extract mathematical expression
            math_pattern = r"(\d+(?:\.\d+)?\s*[+\-*/]\s*\d+(?:\.\d+)?)"
            matches = re.findall(math_pattern, message)
            
            if matches:
                expression = matches[0].replace(" ", "")
                
                # Safe evaluation of simple math expressions
                allowed_chars = set("0123456789+-*/.() ")
                if all(c in allowed_chars for c in expression):
                    try:
                        result = eval(expression)  # Safe for simple math
                        
                        responses = {
                            "EN": f"The result of {expression} is {result}",
                            "FR": f"Le résultat de {expression} est {result}",
                            "PT": f"O resultado de {expression} é {result}"
                        }
                        
                        return {
                            "response": responses.get(language, responses["EN"]),
                            "result": result,
                            "expression": expression,
                            "task_type": "calculation"
                        }
                        
                    except Exception:
                        pass
            
            # Fall back to AI for complex math
            return await self._handle_with_ai(
                message,
                language,
                "You are a helpful math assistant. Solve this mathematical problem step by step."
            )
            
        except Exception as e:
            self.logger.warning("Calculation handling failed", error=str(e))
            
            error_responses = {
                "EN": "I couldn't process that calculation. Please try a simpler expression.",
                "FR": "Je n'ai pas pu traiter ce calcul. Veuillez essayer une expression plus simple.",
                "PT": "Não consegui processar esse cálculo. Tente uma expressão mais simples."
            }
            
            return {
                "response": error_responses.get(language, error_responses["EN"]),
                "task_type": "calculation_error"
            }
    
    async def _handle_translation(
        self,
        message: str,
        language: SupportedLanguage
    ) -> Dict[str, Any]:
        """Handle translation requests."""
        try:
            system_prompts = {
                "EN": "You are a helpful translation assistant. Provide accurate translations and explain the meaning when helpful.",
                "FR": "Vous êtes un assistant de traduction utile. Fournissez des traductions précises et expliquez le sens quand c'est utile.",
                "PT": "Você é um assistente de tradução útil. Forneça traduções precisas e explique o significado quando útil."
            }
            
            result = await self._handle_with_ai(
                message,
                language,
                system_prompts.get(language, system_prompts["EN"])
            )
            
            result["task_type"] = "translation"
            return result
            
        except Exception as e:
            self.logger.warning("Translation handling failed", error=str(e))
            
            error_responses = {
                "EN": "I couldn't process that translation request. Please be more specific about what you'd like to translate.",
                "FR": "Je n'ai pas pu traiter cette demande de traduction. Veuillez être plus précis sur ce que vous voulez traduire.",
                "PT": "Não consegui processar essa solicitação de tradução. Seja mais específico sobre o que você gostaria de traduzir."
            }
            
            return {
                "response": error_responses.get(language, error_responses["EN"]),
                "task_type": "translation_error"
            }
    
    async def _handle_help(
        self,
        message: str,
        language: SupportedLanguage
    ) -> Dict[str, Any]:
        """Handle help requests."""
        help_responses = {
            "EN": """I'm Umbra, your AI assistant! I can help you with:

🏦 **Finance**: Upload invoices/receipts for OCR processing, generate financial reports
🖥️ **VPS Management**: Monitor server status, execute commands, manage Docker containers
👥 **Business**: Manage clients, delegate tasks to other modules
⚙️ **Workflows**: Create n8n automation workflows
🎨 **Media**: Generate images, videos, audio content
🔢 **Simple Tasks**: Math calculations, translations, general questions

Just tell me what you'd like to do, and I'll route your request to the right module or handle it directly!""",
            
            "FR": """Je suis Umbra, votre assistant IA ! Je peux vous aider avec :

🏦 **Finance** : Téléchargez factures/reçus pour traitement OCR, générez des rapports financiers
🖥️ **Gestion VPS** : Surveillez l'état du serveur, exécutez des commandes, gérez les conteneurs Docker
👥 **Business** : Gérez les clients, déléguez des tâches aux autres modules
⚙️ **Workflows** : Créez des workflows d'automatisation n8n
🎨 **Médias** : Générez des images, vidéos, contenu audio
🔢 **Tâches simples** : Calculs mathématiques, traductions, questions générales

Dites-moi simplement ce que vous voulez faire, et je dirigerai votre demande vers le bon module ou la traiterai directement !""",
            
            "PT": """Eu sou Umbra, seu assistente de IA! Posso ajudar você com:

🏦 **Finanças**: Envie faturas/recibos para processamento OCR, gere relatórios financeiros
🖥️ **Gerenciamento VPS**: Monitore status do servidor, execute comandos, gerencie containers Docker
👥 **Negócios**: Gerencie clientes, delegue tarefas para outros módulos
⚙️ **Workflows**: Crie workflows de automação n8n
🎨 **Mídia**: Gere imagens, vídeos, conteúdo de áudio
🔢 **Tarefas simples**: Cálculos matemáticos, traduções, perguntas gerais

Apenas me diga o que você gostaria de fazer, e eu direcionarei sua solicitação para o módulo certo ou lidarei com ela diretamente!"""
        }
        
        return {
            "response": help_responses.get(language, help_responses["EN"]),
            "task_type": "help"
        }
    
    async def _handle_clarification(
        self,
        message: str,
        language: SupportedLanguage
    ) -> Dict[str, Any]:
        """Handle clarification requests."""
        clarification_responses = {
            "EN": "I need more information to help you properly. Could you please be more specific about what you'd like to do?",
            "FR": "J'ai besoin de plus d'informations pour vous aider correctement. Pourriez-vous être plus précis sur ce que vous voulez faire ?",
            "PT": "Preciso de mais informações para ajudá-lo adequadamente. Você poderia ser mais específico sobre o que gostaria de fazer?"
        }
        
        return {
            "response": clarification_responses.get(language, clarification_responses["EN"]),
            "task_type": "clarification"
        }
    
    async def _handle_general_query(
        self,
        message: str,
        language: SupportedLanguage
    ) -> Dict[str, Any]:
        """Handle general queries with AI."""
        try:
            system_prompts = {
                "EN": "You are Umbra, a helpful AI assistant. Provide concise, helpful responses. If the request seems complex or specialized, suggest using specific modules (finance, VPS management, business, workflows, or media generation).",
                "FR": "Vous êtes Umbra, un assistant IA utile. Fournissez des réponses concises et utiles. Si la demande semble complexe ou spécialisée, suggérez d'utiliser des modules spécifiques (finance, gestion VPS, business, workflows, ou génération de médias).",
                "PT": "Você é Umbra, um assistente de IA útil. Forneça respostas concisas e úteis. Se a solicitação parecer complexa ou especializada, sugira usar módulos específicos (finanças, gerenciamento VPS, negócios, workflows, ou geração de mídia)."
            }
            
            result = await self._handle_with_ai(
                message,
                language,
                system_prompts.get(language, system_prompts["EN"])
            )
            
            result["task_type"] = "general_query"
            return result
            
        except Exception as e:
            self.logger.warning("General query handling failed", error=str(e))
            
            error_responses = {
                "EN": "I'm having trouble processing your request right now. Please try again or be more specific.",
                "FR": "J'ai des difficultés à traiter votre demande en ce moment. Veuillez réessayer ou être plus précis.",
                "PT": "Estou tendo problemas para processar sua solicitação agora. Tente novamente ou seja mais específico."
            }
            
            return {
                "response": error_responses.get(language, error_responses["EN"]),
                "task_type": "general_error"
            }
    
    async def _handle_with_ai(
        self,
        message: str,
        language: SupportedLanguage,
        system_prompt: str
    ) -> Dict[str, Any]:
        """Handle request using AI."""
        try:
            messages = [
                {"role": "system", "content": system_prompt},
                {"role": "user", "content": message}
            ]
            
            response = await self.openrouter_client.chat_completion(
                model="openai/gpt-3.5-turbo",
                messages=messages,
                temperature=0.7,
                max_tokens=500
            )
            
            ai_response = response.choices[0].message.content
            
            return {
                "response": ai_response,
                "task_type": "ai_response",
                "model_used": "gpt-3.5-turbo",
                "tokens_used": response.usage.total_tokens
            }
            
        except Exception as e:
            self.logger.error("AI handling failed", error=str(e))
            raise